In [1]:
# importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
df_prods = pd.read_csv(r'C:\Users\gkgiz\Desktop\CF\Achievement 4\09 - 2024 Instacart Basket Analysis\02 Data\Original Data\products.csv', index_col = False)

In [3]:
df_ords = pd.read_csv(r'C:\Users\gkgiz\Desktop\CF\Achievement 4\09 - 2024 Instacart Basket Analysis\02 Data\Prepared Data\orders_wrangled.csv', index_col = False)

In [4]:
# create a new dataframe
df_test = pd.DataFrame()

In [5]:
# create a mixed-type column
df_test['mix'] = ['a', 'b', 1, True]

In [6]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)
      

mix


C:\Users\gkgiz\AppData\Local\Temp\ipykernel_9336\2719724061.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [7]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [8]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [9]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [10]:
df_prods.shape

(49693, 5)

In [11]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [12]:
df_prods_clean.shape

(49677, 5)

In [13]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [14]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [15]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [16]:
df_prods_clean_no_dups.shape

(49672, 5)

# Task 4.5

In [18]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [19]:
# Disable scientific notation
pd.set_option('display.float_format', '{:.6f}'.format)

In [20]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0           0   2539329        1    prior             1          2   
1           1   2398795        1    prior             2          3   
2           2    473747        1    prior             3          3   
3           3   2254736        1    prior             4          4   
4           4    431534        1    prior             5          4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7               15.000000  
2                 12               21.000000  
3                  7               29.000000  
4                 15               28.000000

Step 2: order number max 100 seems a bit suspicious. days since prior order has some missing values, since it has a different count than the other columns. Unnamed 0 can probably be dropped.

In [22]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)
  else :
    print ("Column " + col + " has no mixed types")

C:\Users\gkgiz\AppData\Local\Temp\ipykernel_9336\2838811324.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


Column Unnamed: 0 has no mixed types


C:\Users\gkgiz\AppData\Local\Temp\ipykernel_9336\2838811324.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


Column order_id has no mixed types


C:\Users\gkgiz\AppData\Local\Temp\ipykernel_9336\2838811324.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


Column user_id has no mixed types


C:\Users\gkgiz\AppData\Local\Temp\ipykernel_9336\2838811324.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


Column eval_set has no mixed types


C:\Users\gkgiz\AppData\Local\Temp\ipykernel_9336\2838811324.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


Column order_number has no mixed types


C:\Users\gkgiz\AppData\Local\Temp\ipykernel_9336\2838811324.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


Column order_dow has no mixed types


C:\Users\gkgiz\AppData\Local\Temp\ipykernel_9336\2838811324.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


Column order_hour_of_day has no mixed types


C:\Users\gkgiz\AppData\Local\Temp\ipykernel_9336\2838811324.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


Column days_since_prior_order has no mixed types


Step 4: There is no mixed-type data in the dataframe.

In [24]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

Step 5: days_since_prior_order column has 206209 missing values. It is highly possible that these values are missing because they refer to the first orders placed by the customers. In order to ensure this, we can check if the frequency of days_since_prior_order and the frequency of the observations where order_number = 1 are the same. If so, then our hypothesis is confirmed and we can create a new column where every observation has a boolean string showing whether it is a new order or not.

In [26]:
# Step 1: Count NaN values in 'days_since_prior_order'
nan_count = df_ords['days_since_prior_order'].isna().sum()

# Step 2: Count occurrences of '1' in 'order_number'
one_count = (df_ords['order_number'] == 1).sum()

# Step 3: Compare the counts
if nan_count == one_count:
    print("The number of NaN values in 'days_since_prior_order' is equal to the number of 1s in 'order_number'.")
else:
    print(f"The counts are different:\nNaN count: {nan_count}\n'1' count: {one_count}")


The number of NaN values in 'days_since_prior_order' is equal to the number of 1s in 'order_number'.


In [27]:
# Create new column 'new_order' where True if order_number is 1, else False
df_ords['new_order'] = df_ords['order_number'] == 1

In [28]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0           0   2539329        1    prior             1          2   
1           1   2398795        1    prior             2          3   
2           2    473747        1    prior             3          3   
3           3   2254736        1    prior             4          4   
4           4    431534        1    prior             5          4   

   order_hour_of_day  days_since_prior_order  new_order  
0                  8                     NaN       True  
1                  7               15.000000      False  
2                 12               21.000000      False  
3                  7               29.000000      False  
4                 15               28.000000      False

In [29]:
df_dups = df_ords[df_ords.duplicated()]
#checking for duplicates

In [30]:
df_dups.shape
#no duplicates found

(0, 9)

In [59]:
df_prods.to_csv('C:/Users/gkgiz/Desktop/CF/Achievement 4/09 - 2024 Instacart Basket Analysis/02 Data/Prepared Data/df_prodds.csv', index=False)

In [31]:
df_ords.to_csv('C:/Users/gkgiz/Desktop/CF/Achievement 4/09 - 2024 Instacart Basket Analysis/02 Data/Prepared Data/df_orids.csv', index=False)